### PR TITLE
[SPARK-16849][SQL] Improve subquery execution by deduplicating the subqueries with the same results

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

Design doc: https://issues.apache.org/jira/secure/attachment/12845016/de-duplicating%20subqueries.pdf

The subqueries in SparkSQL will be run even they have the same physical plan and output same results. We should be able to deduplicate these subqueries which are referred in a query for many times.

This change is motivated by the attempt to improve the performance of CTE subqueries. However, it is not limited to CTE but can be applied to other subqueries in general. Note that because CTE provides the scope information, we can better utilize it to find common subqueries among the whole query plan.

For the other subqueries, we rely on `sameResult` to decide if two subqueries are the same common subquery. As we know, `sameResult` doesn't guarantee to determine two given plans will produce the same results, so it is possibly two non-CTE subqueries which produce the same results are not considered as common subquery.
### Compared to existing result reusing rule ReuseExchange

The planning rule `ReuseExchange` can find out duplicated exchanges in the spark plan and reuse the same exchange for all the references. So we already have it to reuse Broadcast and Shuffle. Why do we need this PR to reuse duplicated subqueries?

The mean we rely on to find the duplicated exchanges is `sameResult`. Quoted from its comments:

> Since its likely undecidable to generally determine if two given plans will produce the same results, it is okay for this function to return false, even if the results are actually the same.

That's said, for some cases (if not most cases), the query plan can't meet the requirement of `sameResult` to reuse Shuffle or Broadcast, even the two plans are actually producing the same results. That is what we observed when benchmarking on TPC-DS queries.

However, CTE subqueries provide the good scope information (i.e., the subquery alias) we can utilize to find duplicated query plans. E.g., for a query like:

WITH cte AS (SELECT \* FROM src) SELECT \* FROM cte a JOIN cte b

It is guaranteed that `a` and `b` are the same, and we can reuse the same result.

Besides, there are operators and situations which don't involve Exchange and we still need to de-duplicate the same subqueries.
### Eliminating SubqueryAlias

Previously, we eliminate all `SubqueryAlias` at the beginning of `Optimizer`. Now it is changed to only eliminate the `SubqueryAlias` which is only referred once in the whole query plan. This is performed by the new `EliminateOneTimeSubqueryAliases`.
### De-duplicate common SubqueryAlias

Another new rule `DedupCommonSubqueries` is introduced to de-duplicate common subqueries before optimization.

Each `SubqueryAlias` has a child logical plan representing the subquery. We generate the executed plan for the logical plan and attach it to a  `CommonSubquery` operator which is just a wrapper for the executed plan.

Note that we will search if any same logical plan exists. So for any distinct subquery logical plan, we will have only one executed plan shared between many `CommonSubquery`. Once the executed plan is executed, its result will be kept and shared between these `CommonSubquery` too.

Note that we will only keep the common subqueries at the highest level in the query. That is said we will strip the common subqueries wrapped in other common subquery. That will simplify the logic to optimize filter/projection pushdown between such nested common subqueries. And the common subquery at the highest level should be the most expensive one to de-duplicate.
### Modifying sameResult of SubqueryAlias

This PR override the default `sameResult` method in `SubqueryAlias`. If two `SubqueryAlias` have the child logical plans generating the same results, the `sameResult` on them should return `true`.
### Optimization regarding common subqueries and operators on them

It is possibly that we have a logical plan including a common subquery like this:

```
Join
  Filter [year = 2000]
    SubqueryAlias a
      ...
  Filter [year = 2001]
    SubqueryAlias b
      ...
```

Optimization rules that push down `Filter` would work for it because the additional `SubqueryAlias`. We know `Filter` and `Project` pushdown can improve the query performance a lot. So we introduce a new optimization rule `OptimizeCommonSubqueries`.

Simply said, the rule finds the common subqueries. For each common subquery, we optimize all subqueries producing the same results with the operator on them. We consider two cases. First, all subqueries have a `Project` on them. Second, all subqueries have a `Filter` on them.

For Project pushdown, we concatenate all project lists and push down it to all subqueries. For Filter pushdown, we concatenate all filtering conditions with a `Or` expression and push down it to all subqueries. For example, the above query plan would look like:

```
Join
  Filter [year = 2000]
    SubqueryAlias a
      Filter [year = 2000 or year = 2001]
      ...
  Filter [year = 2001]
    SubqueryAlias b
      Filter [year = 2000 or year = 2001]
      ...
```

Note that the two subqueries have the same filter operator pushed down. So they are still common subqueries and the results can be reused.
## How was this patch tested?

Jenkins tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan @hvanhovell Can you take a look? Thanks!

**Comment by viirya** [comment]
ping @cloud-fan @hvanhovell Can you look at this if it is making sense for you? Thanks.

**Comment by viirya** [comment]
ping @cloud-fan @hvanhovell @liancheng Could you review this change? Thanks.

**Comment by gatorsmile** [comment]
Will the deduplication logics on conflicting attributes in Analyzer affect your solution? 

https://github.com/apache/spark/blob/06f5dc841517e7156f5f445655d97ba541ebbd7e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L478-L482

**Comment by viirya** [comment]
@gatorsmile I think it doesn't affect this. This change is happened after analysis.

**Comment by gatorsmile** [comment]
Will it affect `sameResult`?

**Comment by viirya** [comment]
I think the deduplicating the conflicting attributes in Analyzer is generating new expression ids. `sameResult` is tolerant of expression id differences.

**Comment by gatorsmile** [comment]
If `sameResult` does not consider the expression id, will it return `true` if two tables have the same name columns? If my understanding is not wrong, it only tolerates `Alias` with the different expression IDs.

**Comment by viirya** [comment]
Yea, but other attribute references will be bound before comparing in `sameResult`. So the expression id differences shouldl be no effect.

**Comment by gatorsmile** [comment]
It is just bound to its own `children`, right? If you have two CTE, their children might have different expression IDs after the deduplication logics on conflicting attributes in Analyzer.

**Comment by viirya** [comment]
Although deduplication creates new expression IDs, these logical plans which have the different expression IDs should have the underlying plan will have the same expression IDs. I have just added a test case for that.

**Comment by viirya** [comment]
also cc @davies

**Comment by hvanhovell** [comment]
@viirya I have taken a look. It looks pretty cool what you are doing. One question: why are you focussing on CTE's only? Shouldn't we look at any part of the query tree?

The other thing is that we already reuse most exchanges. What does this PR add performance wise?

**Comment by viirya** [comment]
@hvanhovell Thanks for looking at this.

The first attempt of this PR is to address all `SubqueryAlias`. But looks like we use `SubqueryAlias` plan in many places. To de-duplicating all of them causes the query plan weird and fails some tests as I remember. So I turn to only address the subqueries in CTE.

I have no cue about the exchange reuse, is it about broadcast exchange? If so, I think not all subqueries are able for broadcasting? Our initial test shows that de-duplicating the CTE subqueries can save half of the execution time for query 64.

BTW, the CTE is the motivation of this PR. We found the duplicate materialization of CTE subqueries in TPC-DS queries (for example, query 64) degrades the performance.

**Comment by hvanhovell** [comment]
We reuse all exchanges (including shuffled ones), see this rule: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/Exchange.scala#L69-L94

The optimizer might be doing us a disservice here, by creating two (or more) similar but not the same plans from a single source plan, which will result in two separately executed plans.

**Comment by viirya** [comment]
@hvanhovell Got it. Thanks for the cue.

I think the thought is the same, to reuse the results of the plans wit same results.

The performance gain of this PR is from the reuse of the spark plan of the subqueries in CTE. It is not only for exchange plan.

For example, the TPC-DS query 64 has two CTEs with long join chain of many tables. If we materialize the CTE subqueries in the query each time it is referred, the performance is bad. With this de-duplication, we can just run the long join once and reuse its results.

**Comment by viirya** [comment]
Originally I use the analyzed logical plans to compare subqueries when finding the subqueries with the same results. But I think that optimized plan might be better because it should have performance advantage. And I think with the same analyzed logical plans, we should have the same optimized plan, is it?

**Comment by gatorsmile** [comment]
@viirya 
> Although deduplication creates new expression IDs, these logical plans which have the different expression IDs should have the underlying plan will have the same expression IDs

This is not guaranteed when the query is complex. The conflicting attributes could be different and then trigger different deduplication.

**Comment by gatorsmile** [comment]
We need a different solution to find the origin of one Attribute. Maybe add another field in the node to record the origin? This might help in your case. 6+ months ago, I talked it with @marmbrus @yhuai about this issue. They might have better ideas now.

**Comment by viirya** [comment]
@gatorsmile Do you have concrete example for that?

**Comment by gatorsmile** [comment]
For example, 
```SQL
WITH cte AS (SELECT * FROM testData a, testData b), 
  cte2 AS (SELECT * FROM cte j1, cte)
  SELECT * FROM cte2 j3, cte j4
```

Could you show the generated plan and explain how we can tell from the plan the subquery execution is reused? Thanks!

**Comment by gatorsmile** [comment]
How about this query?

```
WITH cte AS (SELECT * FROM testData a, testData b), 
  cte2 AS (SELECT * FROM testData a)
  SELECT * FROM cte2 j3, cte
```

We still can see `CommonSubquery` in the plan?

**Comment by viirya** [comment]
@gatorsmile I've made some changes. I will update this soon.

The optimized plan for the query is:

    Join Inner
    :- Join Inner
    :  :- CommonSubquery [a#226, b#227, a#245, b#246]
    :  :  :  +- BroadcastNestedLoopJoin BuildRight, Inner, true
    :  :  :     :- LocalTableScan [a#226, b#227]
    :  :  :     +- BroadcastExchange IdentityBroadcastMode
    :  :  :        +- LocalTableScan [a#245, b#246]
    :  +- CommonSubquery [a#247, b#248, a#251, b#252]
    :     :  +- BroadcastNestedLoopJoin BuildRight, Inner, true
    :     :     :- LocalTableScan [a#226, b#227]
    :     :     +- BroadcastExchange IdentityBroadcastMode
    :     :        +- LocalTableScan [a#245, b#246]
    +- CommonSubquery [a#253, b#254, a#257, b#258]
       :  +- BroadcastNestedLoopJoin BuildRight, Inner, true
       :     :- LocalTableScan [a#226, b#227]
       :     +- BroadcastExchange IdentityBroadcastMode
       :        +- LocalTableScan [a#245, b#246]

**Comment by viirya** [comment]
@dilipbiswal yea. Only when the execution result of CommonSubquery is asked, it will be materialized then.

**Comment by viirya** [comment]
@gatorsmile Your second query example have two different CTE subqueries. And no one is reused in the query plan. So we will not see CommonSubquery. The optimized plan of this query looks like:

    Join Inner
    :- LocalRelation [a#226, b#227]
    +- Join Inner
       :- LocalRelation [a#1509, b#1510]
       +- LocalRelation [a#1507, b#1508]

**Comment by viirya** [comment]
@hvanhovell By fixing some bugs in latest code changes, this PR now can be general to all subqueries, instead of just CTE subqueries. Please suggest and review when you can. Thanks!

**Comment by davies** [comment]
@viirya Had a quick look on this one, It seems that we are trying to make the logical plans looks better, but does not help much for actual physical plan or runtime. For example, for self join, the plan had to executed twice (because of the current poll model) without materialization. If they have Shuffle or Broadcast, we can already re-use them. The only benefit I can imagine is that this could help to prevent push different filter on both side which prevent them could be reused (the result will be different). So the question is that should we push down this filter or not? It's not obvious to me, it could be good or bad based on use cases.

So I don't see much benefits from this PR (not a trivial change). Maybe I misunderstand something, please correct me by show some real use cases, thanks!

**Comment by viirya** [comment]
@davies Thanks for comment.

This is proposed to reuse common subquery results in the query plan. For example,

WITH cte as (SELECT * FROM src) SELECT * FROM cte a JOIN cte b

In the above query, the subquery cte will be materialized twice. With this PR, we just execute cte once. In the benchmark of TPC-DS query64, we have cut half of the running time with this PR.

Simply said, we find the common subqueries in the query plan. For each distinct subquery, we have only one SparkPlan for it and attach it to the subqueries which produce the same results. Once the first subquery among them is executed, the RDD is kept and reused by other subqueries.

We would continue to benchmark TPC-DS queries with this PR. Once the results are done, I will post here.

Please let me know if I need to explain more in details.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
@davies Besides, it is usually the query plan can't meet the requirement to reuse Shuffle or Broadcast, as we observed from the query 64. Subquery, especially CTE, provides the scope information we can use to better find common reusable sub query plan.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
Jenkins seems not working?

**Comment by hvanhovell** [comment]
@viirya I am still trying to figure out the added value of this PR. Here is my main question/concern: Spark pipelines operators in a single stage. Results are only materialized at the end of a stage (Exchanging/Collect/Writing data). So what is the point of reusing a plan if it not materialized?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, Statistics}
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.util.Utils
+
+private[sql] case class CommonSubquery(
+    output: Seq[Attribute],
+    @transient child: SparkPlan)(
+    @transient val logicalChild: LogicalPlan,
+    private[sql] val _statistics: Statistics,
+    @transient private[sql] var _computedOutput: RDD[InternalRow] = null)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala]: This does not guarantee that the result is materialized.

**Comment by viirya** [comment]
@hvanhovell Let me try to explain this with an example.

    WITH cte AS (SELECT * FROM src) SELECT * FROM cte a JOIN cte b

In above query, the common subquery `cte` will be executed twice. We find such common subqueries and wrap the executed plan of it into `CommonSubquery` node. These common subqueries which have the same results, will share the same executed plan and the same variable of computed results.

In planning, we create `CommonSubqueryExec` for `CommonSubquery`. When `CommonSubqueryExec.doExecute` is called to materialized the results, we delegate to the executed plan wrapped in `CommonSubquery` and keep its results. As all common subqueries share the same executed plan and the variable of computed results, the later calling on `CommonSubqueryExec.doExecute` can directly take the computed results.

We benchmark this patch on TPC-DS queries and see significant improvement on many queries which use CTE subqueries. We are trying to solve some filter pushdown issues and improve it further.

Please let me know if it is clear for you.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala (hunk 1 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala]: oh. I make a mistake here... I will fix it.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala]: I was thinking that `_computedOutput` will not be kept for all `CommonSubquery` sharing it. But it is not true. It will. So I think it is no problem.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala]: `_computedOutput` is used to keep the computed results across the common subqueries. Once the `CommonSubqueryExec` of any subquery among them is asked for the results, we will execute the executed plan and record the results in `computedOutput`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -89,6 +90,8 @@ abstract class Optimizer(sessionCatalog: SessionCatalog, conf: CatalystConf)
       CombineFilters,
       CombineLimits,
       CombineUnions,
+      // Pushdown Filters again after combination
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This change is submitted at #14912. Because it prevents this PR pushdown some predicates, for the easy of test, I also include that change here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala (hunk 1 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala]: @hvanhovell Rethinking about this, it is incorrect indeed. I change it and use a helper class to do rdd materialization.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: #14912 is changed. But the changed solution is more complicated so I don't want to include it here. So I keep this change and wait for #14912 to be merged first.

**Comment by viirya** [comment]
@davies @hvanhovell When the physical plan of common subqueries is executed at the first time, we immediately call its `count()`. Doesn't this call make it materialized?

**Comment by viirya** [comment]
There is an example which current shuffle exchange reuse can't work, TPC-DS query 47. The query looks like:

    with v1 as(
    select i_category, i_brand,
           s_store_name, s_company_name,
           d_year, d_moy,
           sum(ss_sales_price) sum_sales,
           avg(sum(ss_sales_price)) over
             (partition by i_category, i_brand,
                        s_store_name, s_company_name, d_year)
             avg_monthly_sales,
           rank() over
             (partition by i_category, i_brand,
                        s_store_name, s_company_name
              order by d_year, d_moy) rn
    from item, store_sales, date_dim, store
    where ss_item_sk = i_item_sk and
          ss_sold_date_sk = d_date_sk and
          ss_store_sk = s_store_sk and
          (
            d_year = 1999 or
            ( d_year = 1999-1 and d_moy =12) or
            ( d_year = 1999+1 and d_moy =1)
          )
    group by i_category, i_brand,
             s_store_name, s_company_name,
             d_year, d_moy),
    v2 as(
    select v1.i_category, v1.i_brand, v1.s_store_name, v1.s_company_name, v1.d_year,
                        v1.d_moy, v1.avg_monthly_sales ,v1.sum_sales, v1_lag.sum_sales psum,
                        v1_lead.sum_sales nsum
    from v1, v1 v1_lag, v1 v1_lead
    where v1.i_category = v1_lag.i_category and
          v1.i_category = v1_lead.i_category and
          v1.i_brand = v1_lag.i_brand and
          v1.i_brand = v1_lead.i_brand and
          v1.s_store_name = v1_lag.s_store_name and
          v1.s_store_name = v1_lead.s_store_name and
          v1.s_company_name = v1_lag.s_company_name and
          v1.s_company_name = v1_lead.s_company_name and
          v1.rn = v1_lag.rn + 1 and
          v1.rn = v1_lead.rn - 1)
    select * from v2
    where  d_year = 1999 and
           avg_monthly_sales > 0 and
           case when avg_monthly_sales > 0 then abs(sum_sales - avg_monthly_sales) / avg_monthly_sales else null end > 0.1
    order by sum_sales - avg_monthly_sales, 3
    limit 100

In the physical plan, the subquery v1 is the common subquery which is used three times (i.e., v1, v1_lag and v1_lead) in the subquery v2. The three v1 subqueries are the child of three shuffle exchange nodes. Because they have different columns (rn, rn + 1, rn - 1) in hash partitioning, the three shuffle exchange nodes are not the same and can't be reused. However, with this patch, we can reuse the common subquery. By doing that, we observe significant improvement in q47.

**Comment by viirya** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import org.apache.spark.util.Utils
+
+private[sql] case class SubqueryExecHelper(executedPlan: SparkPlan) {
+  private var _computedOutput: RDD[InternalRow] = null
+
+  def computeOrGetResult(): RDD[InternalRow] = this.synchronized {
+    if (_computedOutput == null) {
+      _computedOutput = executedPlan.execute().mapPartitionsInternal { rowIterator =>
+        rowIterator.map(_.copy())
+      }.cache()
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala]: We need to cache explicitly the result of common subquery. Becasue in RDD materialization implementation, if a RDD is not marked to cache, even it is materialization before, the later call to its iterator will invoke computation again.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 4]
```diff
@@ -268,3 +269,35 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
     case operator: SparkPlan => ensureDistributionAndOrdering(operator)
   }
 }
+
+
+/**
+ * Pushs down [[ShuffleExchange]] on the top of [[CommonSubqueryExec]] to make them reusable.
+ */
+case object PushDownShuffleExchange extends Rule[SparkPlan] {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: The common pattern of common subquery in TPC-DS, is joining two or more the same subquery. The joining is `SortMergeJoin` and `ShuffleExchange` is executed for all the subqueries after some operators (e.g., `Project`, `Filter`).

E.g.,

      :     :- *SortMergeJoin [customer_id#898], [customer_id#912], Inner
      :     :  :- *Sort [customer_id#898 ASC], false, 0
      :     :  :  +- Exchange hashpartitioning(customer_id#898, 200)
      :     :  :     +- *Project [customer_id#898, year_total#902]
      :     :  :        +- *Filter (((((isnotnull(year#901) && isnotnull(year_total#902)) && (sale_type#903 = s)) && (year#901 = 2001)) && (year_total#902 > 0.0)) && isnotnull(customer_id#898))
      :     :  :           +- CommonSubquery [customer_id#898, customer_first_name#899, customer_last_name#900, year#901, year_total#902, sale_type#903]

This rule is to pushdown `ShuffleExchange` on the top of `CommonSubqueryExec`. So other `ShuffleExchange`s can be transformed to `ReusedExchange` later. Then we can reuse the exchange.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala (hunk 3 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery/CommonSubquery.scala]: I revert this change because don't see improvement with it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala (hunk 4 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: I revert this change because don't see improvement with it.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
@hvanhovell @davies I rethink this PR in recent days. The changes includes some hacky change and are too big to review. I would like to separate it to individual small PRs which can be reviewed separately.

**Comment by davies** [comment]
@viirya It's true that this PR could help the performance for Q64, but i'm not convinced that that case is common enough to worth more than 500+ LOC changes. Could you try run TPC-DS Q64 by disable some predicate push down (for example, disable InferFiltersFromConstraints)?

**Comment by viirya** [comment]
@davies Thanks for comment.

In our initial benchmark of the TPC-DS queries (totally 13) using CTE, this PR helps about half (6) of them, 5 queries are not affected, 2 queries are regressed. I might say it would help CTE queries at most cases according to the results.

I agree that 500+ LOC changes looks a bit big for this improvement. I would like to refactor and tailor part of the changes and break it down to small pieces for you to review. I wish I can reduce the LOC changes at the end.

I would like to run Q64 by disabling the push down, may I ask the purpose for it?

**Comment by rxin** [comment]
Adding an explicit cache call btw -- can actually slow things down due to bad memory management.

**Comment by viirya** [comment]
@rxin yeah, as I tried adding explicit cache call doesn't improve it. So I remove it then.

**Comment by viirya** [comment]
Revisit this by rebasing with master.

BTW, in 500+ LOC changes, actually there are 200+ LOC changes are test cases.

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
After reading the PR description, I think this improvement is quite complex and worth a design doc. We should explain how the subquery execution works now and how you are going to change it.

**Comment by viirya** [comment]
@cloud-fan Thanks for comment. I agreed that. I will prepare a design doc soon.

**Comment by davies** [comment]
@viirya For duplicated CTE, without some optimization (pushing down different predicates in different positions), the physical plan should be identical. So I'm wondering some aggressive pushing down cause the problem for some queries (IsNotNull(xxx)). This is the reason I asked that.

**Comment by viirya** [comment]
@davies It is true that pushing down different predicates results in different CTE logical/physical plans. I spend some LOC changes in this to address that cases, i.e., preparing a disjunctive predicate for duplicated CTE with different predicates.

For Q64, a disjunctive predicate will be pushed down too. I am not sure what the problem is you mentioned.

Let me try to get and show the pushed down predicate.

**Comment by viirya** [comment]
I would like to share some numbers I ran this on my local cluster (5 nodes, on yarn, 8GB).

after this:

q2: 9955 ms
q11: 23490 ms
q39a: 11226 ms
q39b: 11303 ms
q47: 21878 ms
q57: 19661 ms
q59: 10803 ms
q65: 9949 ms
q74: 20290 ms
q75: 21561 ms

before this:

q2: 9414 ms
q11: 29764 ms
q39a: 12961 ms
q39b: 11578 ms
q47: 46424 ms
q57: 34798 ms
q59: 10420 ms
q65: 10259 ms
q74: 29574 ms
q75: 27767 ms

q64 causes out-of-memory on the cluster. These queries are CTE queries.

The data is generated with spark-sql-perf using scaleFactor = 1 to make these queries working on my cluster.

**Comment by cloud-fan** [comment]
can you provide a typical query that can benefit from this PR? And also show why it's slow without your PR. Thanks!

**Comment by viirya** [comment]
@cloud-fan ok. I will do it.

**Comment by viirya** [comment]
@rxin has pointed out:
> Adding an explicit cache call btw -- can actually slow things down due to bad memory management.

In my experiment, I found:

According to my test in local cluster, without the cache call, I can't see actual speed-up. Only with cache call, I can see improvement.

However the effective caching for rows needs additional `copy` call which adds costs and reduces the improvement.

**Comment by viirya** [comment]
Q47 is a typical query that can benefit from this PR:

    with v1 as(
    select i_category, i_brand,
           s_store_name, s_company_name,
           d_year, d_moy,
           sum(ss_sales_price) sum_sales,
           avg(sum(ss_sales_price)) over
             (partition by i_category, i_brand,
                        s_store_name, s_company_name, d_year)
             avg_monthly_sales,
           rank() over
             (partition by i_category, i_brand,
                        s_store_name, s_company_name
              order by d_year, d_moy) rn
    from item, store_sales, date_dim, store
    where ss_item_sk = i_item_sk and
          ss_sold_date_sk = d_date_sk and
          ss_store_sk = s_store_sk and
          (
            d_year = 1999 or
            ( d_year = 1999-1 and d_moy =12) or
            ( d_year = 1999+1 and d_moy =1)
    group by i_category, i_brand,
             s_store_name, s_company_name,
             d_year, d_moy),
    v2 as(
    select v1.i_category, v1.i_brand, v1.s_store_name, v1.s_company_name, v1.d_year,
           v1.d_moy, v1.avg_monthly_sales ,v1.sum_sales, v1_lag.sum_sales psum,
           v1_lead.sum_sales nsum
    from v1, v1 v1_lag, v1 v1_lead
    where v1.i_category = v1_lag.i_category and
          v1.i_category = v1_lead.i_category and
          v1.i_brand = v1_lag.i_brand and
          v1.i_brand = v1_lead.i_brand and
          v1.s_store_name = v1_lag.s_store_name and
          v1.s_store_name = v1_lead.s_store_name and
          v1.s_company_name = v1_lag.s_company_name and
          v1.s_company_name = v1_lead.s_company_name and
          v1.rn = v1_lag.rn + 1 and
          v1.rn = v1_lead.rn - 1)
    select * from v2
    where  d_year = 1999 and
           avg_monthly_sales > 0 and
           case when avg_monthly_sales > 0 then abs(sum_sales - avg_monthly_sales) / avg_monthly_sales else null end > 0.1
    order by sum_sales - avg_monthly_sales, 3
    limit 100

It has 2 CTEs, `v1` and `v2`. `v2` joins three duplicated `v1` plans. There are no disjunctive predicates needed to be pushed down to those `v1` plans. So the three duplicated `v1` plans have the same output data in the end. The physical plan of `v1` looks complicated.

Obviously without this PR we will run three physical plans of `v1`. This PR runs the physical plan of `v1` for only one time and cache its output data.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -1167,3 +1173,211 @@ object RemoveRepetitionFromGroupExpressions extends Rule[LogicalPlan] {
...
+
+  private def optimizeFilterWithSubqueries(
+      plan: LogicalPlan,
+      keyPlan: LogicalPlan,
+      subqueries: ArrayBuffer[LogicalPlan]): LogicalPlan = {
+    var pushdownConds = splitConjunctivePredicates(subqueries(0).asInstanceOf[Filter].condition)
+    subqueries.tail.foreach {
+      case Filter(otherCond, child) =>
+        val rewrites = buildRewrites(child, subqueries(0).asInstanceOf[Filter].child)
+        // We can't simply push down all conditions from other Filter by concatenating them with
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This part has been extracted out as #15558 and can be removed if that PR is merged.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
Because the actual improvement of this de-duplication depends on the complexity of disjunctive predicate pushdown and CTE subquery, if we can't have general rule to decide whether de-duplicate or not, I think can we have a config (default off) to enable/disable this subquery de-duplication.
